torch/_export/serde/serialize.py

```diff
@@ -633,15 +632,150 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+        self.model_opset_version = model_opset_version if model_opset_version else {}
+        # key is version number, value is a list of upgraders, keyed on op name
+        self.upgraders: List[Tuple[str, str]] = self._parse_upgraders(op_upgraders)
+
+        self.upgrader_passes: List[GraphModuleOpUpgrader.UpgraderPass] = self._populate_passes()
+
+    def _parse_upgraders(self, op_upgraders: Optional[Dict[str, Tuple[str, str]]] = None) -> List[Tuple[str, str]]:
+        """reorder op_upgraders by version number."""
+        # TODO(larryliu0820): Add support for custom ops
+        if not op_upgraders or "aten" not in self.model_opset_version or "aten" not in self.compiler_opset_version:
+            return []
+        model_ver = self.model_opset_version["aten"]
+        curr_ver = self.compiler_opset_version["aten"]
+
+        def get_target_version(versioned_upgrader_name: str) -> int:
+            """div_Scalar_0_3 is the name of the upgrader, meaning it applies to opset version 0 to 3 and is
+            upgrading to version 4."""
+            return int(versioned_upgrader_name.split('_')[-1]) + 1
+
+        versioned_upgraders: Dict[int, Tuple[str, str]] = {get_target_version(name): v for name, v in op_upgraders.items()}
+        target_upgraders: List[Tuple[str, str]] = []
+        for ver in range(model_ver, curr_ver + 1):
+            if ver in versioned_upgraders:
+                target_upgraders.append(versioned_upgraders[ver])
+
+        return target_upgraders
+
+    def _populate_passes(self) -> List[UpgraderPass]:
+        """Given a list of upgraders, loop through it from lower version to higher version and create passes for all
+        upgraders. se torch.Library API to register old ops. Op name will be
+        <name>_<valid_from_ver>_<valid_till_ver>. Register upgarders as CompositeImplicitAutograd kernels. For example:
+
+        lib = Library("aten", "FRAGMENT")
+        lib.define(old_schema)
+
+        impl_lib = Library("aten", "IMPL")
+        impl_lib.impl("div__Scalar_0_3", div__Scalar_0_3, "CompositeImplicitAutograd")
+        """
+
+        upgrader_passes = []
+
+        def register_old_op(name: str, schema: str, impl_str: str):
+            """Registers an old version operator using impl_name as old op name."""
+            lib.define(schema)
+            exec(impl_str)
+            impl_lib.impl(name, locals()[name], "CompositeImplicitAutograd")
+
+        for (schema, upgrader_str) in self.upgraders:
+            upgrader_name = upgrader_str.split('(')[0].split(' ')[-1]
+            op_name = schema.split('(')[0].split("::")[-1]
+            schema = schema.replace(op_name, upgrader_name)
+            try:
+                register_old_op(name=upgrader_name, schema=schema, impl_str=upgrader_str)
+            except RuntimeError as e:
+                if "with the same name and overload name multiple times" in str(e):
```

Comment:
lol is it possible to have a special error type so we don't have match error message specifically?

Reply:
yeah this is part of torch.library.Library API. I can update it in later PRs.